kong/pdk/tracing.lua

```diff
@@ -0,0 +1,452 @@
+
+  local sampled = parent_span and parent_span.sampled
+    or options.sampled
+    or tracer.sampler(trace_id) == FLAG_SAMPLED_AND_RECORDING
+
+  if not sampled then
+    return noop_span
+  end
+
+  -- avoid reallocate
+  -- we will release the span table at the end of log phase
+  local span = tablepool.fetch("KONG_SPAN", 0, 12)
+  -- cache tracer ref, to get hooks / span processer
+  -- tracer ref will not be cleared when the span table released
+  span.tracer = tracer
+
+  span.name = name
+  span.trace_id = trace_id
+  span.span_id = generate_span_id()
+  span.parent_id = parent_span and parent_span.span_id
+      or options.parent_id
+
+  -- specify span start time manually
+  span.start_time_ns = options.start_time_ns or ffi_time_unix_nano()
+  span.kind = options.kind or SPAN_KIND.INTERNAL
+  span.attributes = options.attributes
+
+  -- indicates whether the span should be reported
+  span.sampled = parent_span and parent_span.sampled
+      or options.sampled
+      or sampled
+
+  -- TODO(mayo): span.is_recording
+
+  span.parent = parent_span
+
+  setmetatable(span, span_mt)
+
+  -- insert the span to ctx
+  local spans = ngx.ctx.KONG_SPANS
+  if not spans then
+    spans = tablepool.fetch("KONG_SPANS", 10, 0)
+    spans[0] = 0
+    ngx.ctx.KONG_SPANS = spans
+  end
+
+  local len = spans[0] + 1
+  spans[len] = span
+  spans[0] = len
+
+  return span
+end
+
+--- Ends a Span
+--
+-- @module Span
+-- @tparam number|nil end_time_ns
+function span_mt:finish(end_time_ns)
+  if self.end_time_ns ~= nil then
+    -- span is finished, and processed already
+    return
+  end
+
+  if end_time_ns ~= nil and type(end_time_ns) ~= "number" then
+    error("invalid span end time", 2)
+  end
+
+  if end_time_ns and end_time_ns < self.start_time_ns then
+    error("invalid span duration", 2)
+  end
+
+  self.end_time_ns = end_time_ns or ffi_time_unix_nano()
+
+  if self.active and self.tracer.active_span() == self then
+    self.tracer.set_active_span(self.parent)
+    self.active = nil
+  end
+end
+
+--- Set an attribute to a Span
+--
+-- @module Span
+-- @tparam string key
+-- @tparam string|number|boolean value
+function span_mt:set_attribute(key, value)
+  if type(key) ~= "string" then
+    error("invalid key", 2)
+  end
+
+  local vtyp = type(value)
+  if vtyp ~= "string" and vtyp ~= "number" and vtyp ~= "boolean" then
+    error("invalid value", 2)
+  end
+
+  if self.attributes == nil then
+    self.attributes = new_tab(0, 4)
+  end
+
+  self.attributes[key] = value
+end
+
+-- Adds an event to a Span
+--
+-- @module Span
+-- @tparam string name Event name
+-- @tparam number|nil time_ns Event timestamp
+-- @tparam number|nil attributes attributes
+function span_mt:add_event(name, attributes, time_ns)
+  if type(name) ~= "string" then
+    error("invalid name", 2)
+  end
+
+  if attributes ~= nil then
+    if type(attributes) ~= "table" then
+      error("invalid attribute", 2)
+    end
+  end
+
+  if self.events == nil then
```

Comment:
Would it make any sense to use tablepool here for `self.events` (since we control its lifecycle), or no?

Reply:
Hmm, seems reasonable to use tablepool for both `events` and `attributes` there.